src/vcTiffConvert.cpp

```diff
@@ -10,157 +10,592 @@
+      pSample->t_double[sampleIndex] = ((double *)(pBuf))[ind];
+    else
+      UD_ERROR_SET(udE_InvalidConfiguration);
+  }
+
+  result = udE_Success;
+epilogue:
+  return result;
+}
 
-class TiffFullImageReader : public TiffPixelReader
+static udError vcTiff_LoadTile(vcTiffConvertData *pData, uint32_t x, uint32_t y)
 {
-public:
+  udError result;
+  uint8_t *pBuf = nullptr;
+  uint32_t px, py;
+
+  UD_ERROR_NULL(pData, udE_InvalidParameter);
+
+  px = x * pData->tile.tileDimensions.x;
+  py = y * pData->tile.tileDimensions.y;
 
-  TiffFullImageReader()
-    : TiffPixelReader()
-    , m_pRaster(nullptr)
-    , m_currentPixel(0)
-    , m_imageDimensions{}
+  pData->tile.currentPixel = 0;
+  pData->tile.currentTileCoords = {x, y};
+
+  if (pData->format.planarconfig == PLANARCONFIG_CONTIG)
+  {
+    if (pData->tile.pRaster == nullptr)
+    {
+      pData->tile.pRaster = (uint8_t *)_TIFFmalloc(TIFFTileSize(pData->pTif));
+      UD_ERROR_NULL(pData->tile.pRaster, udE_MemoryAllocationFailure);
+    }
+    UD_ERROR_IF(TIFFReadTile(pData->pTif, pData->tile.pRaster, px, py, 0, 0) == 0, udE_ReadFailure);
+  }
+  else
   {
+    if (pData->tile.pRaster == nullptr)
+    {
+      pData->tile.pRaster = (uint8_t *)_TIFFmalloc(TIFFTileSize(pData->pTif) * udMin(pData->format.samplesPerPixel, MAX_SAMPLES));
+      UD_ERROR_NULL(pData->tile.pRaster, udE_MemoryAllocationFailure);
+    }
+
+    size_t pixelCount = (size_t)pData->tile.tileDimensions.x * pData->tile.tileDimensions.y;
+    pBuf = (uint8 *)_TIFFmalloc(TIFFTileSize(pData->pTif)); // TODO could allocated this once per image
+    UD_ERROR_NULL(pBuf, udE_MemoryAllocationFailure);
 
+    for (uint16 i = 0; i < udMin(pData->format.samplesPerPixel, MAX_SAMPLES); ++i)
+    {
+      UD_ERROR_IF(TIFFReadTile(pData->pTif, pBuf, px, py, 0, i) == 0, udE_ReadFailure);
+      vcTiff_PackSamples(pData->tile.pRaster, pBuf, (uint32_t)pixelCount, pData->format.bitsPerSample, pData->format.samplesPerPixel, i);
+    }
   }
 
-  ~TiffFullImageReader()
+  result = udE_Success;
+epilogue:
+  _TIFFfree(pBuf);
+  return result;
+}
+
+static udError vcTiffImageReaderInit_Tile(vcTiffConvertData *pData)
+{
+  udError result;
+
+  if (pData == nullptr)
+    UD_ERROR_SET(udE_InvalidParameter);
+
+  pData->tile = {};
+
+  UD_ERROR_IF(TIFFGetField(pData->pTif, TIFFTAG_IMAGEWIDTH, &pData->imageDimensions.x) != 1, udE_ReadFailure);
+  UD_ERROR_IF(TIFFGetField(pData->pTif, TIFFTAG_IMAGELENGTH, &pData->imageDimensions.y) != 1, udE_ReadFailure);
+  UD_ERROR_IF(TIFFGetField(pData->pTif, TIFFTAG_TILEWIDTH, &pData->tile.tileDimensions.x) != 1, udE_ReadFailure);
+  UD_ERROR_IF(TIFFGetField(pData->pTif, TIFFTAG_TILELENGTH, &pData->tile.tileDimensions.y) != 1, udE_ReadFailure);
+
+  UD_ERROR_CHECK(vcTiff_LoadTile(pData, 0, 0));
+
+  result = udE_Success;
+epilogue:
+  return result;
+}
+
+static void vcTiffImageReaderDestroy_Tile(vcTiffConvertData *pData)
+{
+  if (pData == nullptr)
+    return;
+
+  _TIFFfree(pData->tile.pRaster);
```

Comment:
Is there a case where this isn't allocated by your code above? Seems like using `udAllocType` would be superior in this case.